src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/PkgControl.java

```diff
@@ -40,28 +47,130 @@
         this.parent = parent;
-        fullPackage = parent.fullPackage + "." + subPkg;
+        if (regex || parent.regex) {
+            // regex gets inherited
+            final String parentRegex = ensureSelfContainedRegex(parent.fullPackage, parent.regex);
+            final String thisRegex = ensureSelfContainedRegex(subPkg, regex);
+            fullPackage = parentRegex + DOT_REGEX + thisRegex;
+            patternForPartialMatch = createPatternForPartialMatch();
+            patternForExactMatch = createPatternForExactMatch();
+            this.regex = true;
+        }
+        else {
+            fullPackage = parent.fullPackage + DOT + subPkg;
+            patternForPartialMatch = null;
+            patternForExactMatch = null;
+            this.regex = false;
+        }
         parent.children.add(this);
     }
 
     /**
+     * Returns a regex that is suitable for concatenation by either converting a plain string into
+     * a regular expression (handling special characters) or by enclosing {@code input} in a
+     * (non-capturing) group if {@code input} already is a regular expression.
+     *
+     * <p>The grouping is strictly necessary if a) {@code input} is a regular expression that b)
+     * contains the alteration character ('|') and if c) the pattern is not already enclosed in a
+     * group - as you see in this example: {@code parent="com|org", child="common|uncommon"} will
+     * result in the pattern {@code "(?:org|com)\.(?common|uncommon)"} what will match
+     * {@code "com.common"}, {@code "com.uncommon"}, {@code "org.common"}, and {@code
+     * "org.uncommon"}. Without the grouping it would be {@code "com|org.common|uncommon"} which
+     * would match {@code "com"}, {@code "org.common"}, and {@code "uncommon"}, which clearly is
+     * undesirable. Adding the group fixes this.
+     *
+     * <p>For simplicity the grouping is added to regular expressions unconditionally.
+     *
+     * @param input the input string.
+     * @param alreadyRegex signals if input already is a regular expression.
+     * @return a regex string.
+     */
+    private static String ensureSelfContainedRegex(final String input, final boolean alreadyRegex) {
+        final String result;
+        if (alreadyRegex) {
+            result = encloseInGroup(input);
+        }
+        else {
+            result = toRegex(input);
+        }
+        return result;
+    }
+
+    /**
+     * Enclose {@code expression} in a (non-capturing) group.
+     * @param expression the input regular expression
+     * @return a grouped pattern.
+     */
+    private static String encloseInGroup(String expression) {
+        return "(?:" + expression + ")";
+    }
+
+    /**
+     * Converts a normal package name into a regex pattern by escaping all
+     * special characters that may occur in a java package name.
+     * @param input the input string.
+     * @return a regex string.
+     */
+    private static String toRegex(String input) {
```

Comment:
please remove this function completely , it is resposibility of user to provide proper regexp.
We do not change user defined configuration. User know what he is doing.